plugins/filter/reproducer_refspec.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+    name: reproducer_refspec
+    short_description: Maps repo information from zuul to a refspec.
+    description:
+        - Maps repo information from zuul to a refspec that should be
+        used for pulling it with git.
+    options:
+        _input:
+            description:
+                - repo information from zuul variables.
+            type: dict
+            required: true
+"""
+
+EXAMPLES = """
+    - name: Get refspec to clone repo from rdo gerrit
+      ansible.builtin.set_fact:
+        change_refspec: >-
+            {{
+              repo | reproducer_refspec
+            }}
+"""
+
+RETURN = """
+  _value:
+    description: The refspec to pull the change.
+    type: str
+    sample:
+      "refs/changes/15/449415/5"
+
+"""
+
+from ansible.errors import AnsibleFilterError, AnsibleFilterTypeError
+
+
+class FilterModule:
+    @classmethod
+    def __map_repo_refspec(cls, repo):
+        if not isinstance(repo, dict):
+            raise AnsibleFilterTypeError(
+                f"reproducer_refspec requires a dict, got {type(repo)}"
+            )
+        if "change" not in repo:
+            # handle case pointing to main/master, e.g. a periodic job
+            return ""
+        change = repo["change"]
+        if "project" not in repo:
+            raise AnsibleFilterError(
+                "repo information does not contain 'project' field"
+            )
+        if "canonical_hostname" not in repo["project"]:
+            raise AnsibleFilterError(
+                "repo information does not contain 'canonical_hostname' field"
+            )
+        hostname = repo["project"]["canonical_hostname"]
+        if "rdoproject" in hostname or "code.eng" in hostname:
+            if "patchset" not in repo:
```

Comment:
trying to think how to tidy up a bit but it is hard as the fields are all different. I was thinking along the lines of moving the checks right at the start:
```
for check in ["change", "project", "patchset"]:
    if check not in repo.keys():
        error
```

Reply:
I'm not sure we can easily do that, as there are many cases where some of those keys are not present and it should not raise an error, e.g, periodic jobs have no `check` field, and changes coming from github like the framework do not have a `patchset` field